platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+
+TODO: Protobuf of data
+
+##### TssVote System Transaction
+
+TODO: Protobuf of data
+
+##### TssShareSignature System Transaction
+
+TODO: Protobuf of data
+
+#### State Datastructures
+
+The state needs to store the relevant ledger id, consensus rosters, key material, and partially constructed ledger
+signatures.
+
+##### Ledger Id
+
+The `ledgerId` is the public ledger key able to verify ledger signatures. It is used by both the application and the
+platform. This value should not change during address book changes and its value does not change unless the network
+goes through a another TSS bootstrap process. Since its value is not expected to change, storing it in a singleton
+Merkle Leaf by itself is appropriate.
+
+TODO: Protobuf of data
+
+##### Consensus Rosters
+
+This proposal extends the roster data format where each `RosterEntry` has a new long term public EC key called a
+`tssEcKey` that is used in the Groth21 algorithm.
+
+TODO: Protobuf of data
+
+##### Key Material (TSS Data Map)
+
+The TSS Data Map is a Map Merkle Leaf that is a combined map of all the key material for all rosters that are
+tracked and the votes by nodes that have validated the key material.
+
+Keys in the TSS Data Map have the following structure: (KeyType, RosterHash, SequenceNumber).
+
+1. KeyType - An enum that is one of TSS_MESSAGE or TSS_VOTE.
+2. RosterHash - The hash of the consensus roster that the message or vote is related to.
+3. SequenceNumber - The order in which the message or vote came to consensus.
+
+The value associated with a TSS_MESSAGE key is the pair (ShareId, TssMessage)
+
+1. ShareId - The id of the share that the message is related to.
+2. TssMessage - the raw bytes of the TssMessage
+
+The value associated with a TSS_VOTE key is a bit vector with the following interpretation
+
+* The order of bits from the least significant bit to the most significant bit corresponds to the numeric order of
+  share ids from least to greatest.
+* If a bit is set to 1, then the TssMessage for the corresponding ShareId was valid and contributed to a successful
+  reconstruction of the ledger id.
+* If a bit is set to 0, then the TssMessage for the corresponding ShareId was either invalid, was not received, or
+  was not used in the reconstruction of the ledger id.
+* If the sum of the bits is greater than the threshold, then the vote is a yes vote in favor of adopting the
+  candidate roster.
+
+Lifecycle Invariants
+
+1. Inserts should only be made for candidate rosters that are going through the keying process.
+2. When a roster is no longer stored in the state, the corresponding key material should be removed.
+3. Every roster that has become active after the TSS bootstrap process must have sufficient key material to be
+   able to recover the ledger id.
+
+TODO: Protobuf of data
+
+##### Partial Ledger Signatures
+
+For a configured number of rounds, share signatures on a message are stored in the state until a threshold number of
```

Comment:
I don't understand why we would need to store partial signatures in the state.

Reply:
Now that we have gotten rid of guaranteed signing, we don't need to.  If a node cannot reconstruct the ledger signature because it missed partial signatures, oh well.   Hopefully at least 1 other node could, and at worst a later block's signature will cover the gap.